Extract performance improvements

dev7355608 opened the pull request:
##### Description of change

- 3d2d89c9d6ad1a6fe6bfb78fec5a2a4dc10d9d8f: `canvas.toBlob` instead of `toDataURL` blocks the main thread for a shorter duration.
- 4a2ff59114c9f015b9c96f7f1c04fd4e53c331c2: Improves performance (~50ms => ~42ms in my test). The `getImageData` is unnessecary.
- 4b2ce461ddd34465d27a138ed245c7eaf9d876ca: Improves performance even more (~42ms => ~34ms in my test).

##### Pre-Merge Checklist

- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

bigtimebuddy commented on packages/canvas-extract/src/CanvasExtract.ts:
```diff
@@ -63,6 +63,20 @@ export class CanvasExtract implements ISystem, IExtract
 
+        if (canvas.toBlob !== undefined)
+        {
+            return await new Promise<string>((resolve, reject) =>
```
The await here is unnecessary and creates an additional promise. We should remove it.

```suggestion
            return new Promise<string>((resolve, reject) =>
```

bigtimebuddy commented on packages/canvas-extract/src/CanvasExtract.ts:
```diff
@@ -71,16 +85,18 @@ export class CanvasExtract implements ISystem, IExtract
             const blob = await canvas.convertToBlob({ type: format, quality });
 
-            return await new Promise<string>((resolve) =>
+            return await new Promise<string>((resolve, reject) =>
```
Same here. 
```suggestion
            return new Promise<string>((resolve, reject) =>
```

bigtimebuddy commented on packages/extract/src/Extract.ts:
```diff
@@ -84,6 +84,20 @@ export class Extract implements ISystem, IExtract
 
+        if (canvas.toBlob !== undefined)
+        {
+            return await new Promise<string>((resolve, reject) =>
```
```suggestion
            return new Promise<string>((resolve, reject) =>
```

bigtimebuddy commented on packages/extract/src/Extract.ts:
```diff
@@ -92,16 +106,18 @@ export class Extract implements ISystem, IExtract
             const blob = await canvas.convertToBlob({ type: format, quality });
 
-            return await new Promise<string>((resolve) =>
+            return await new Promise<string>((resolve, reject) =>
```
```suggestion
            return new Promise<string>((resolve, reject) =>
```

bigtimebuddy reviewed (approved):
Nice one!

dev7355608 commented:
I noticed my y-flipping was flawed: I forgot about the 4 bytes per pixel. There was no test that could have caught this error: so I added a bunch of tests that checks that looks that the extracted pixels/canvas.

I also replaced `postArrayDivide` with a much faster version (more that 3 times faster in my test).

dev7355608 commented:
I added test for `CanvasExtract` too. Notice that `Extract.pixels` from a renderer does not y-flip, while `CanvasExtract.pixels` returns y-flipped results. Is this a bug? Wouldn't it make more sense if `Extract.pixels` y-flipped?

Another difference that bothers me a little bit is that `CanvasExtract.pixels` returns `Uint8ClampedArray` and `Extract.pixels` returns `Uint8Array`.

bigtimebuddy reviewed (commented):
Can you fix the strict type issues?
